File: Manipulator3D_C_Sharp/src/robot/RobotArm.cs
using System;
using System.Numerics;

namespace Manipulator3D.robot;

public struct LinkParams
{
    public float length_m;
    public float mass_kg;

    // Uniform rod approximations (about center, axis perpendicular to rod).
    // I_cm   = (1/12) m L^2
    // I_joint= (1/3)  m L^2
    public float inertia_cm;
    public float inertia_joint;

    public void RecomputeInertia()
    {
        inertia_cm = (1.0f / 12.0f) * mass_kg * length_m * length_m;
        inertia_joint = (1.0f / 3.0f) * mass_kg * length_m * length_m;
    }
}

public struct JointAngles
{
    public float q0_yaw;   // rad
    public float q1_pitch; // rad (relative to XY plane)
    public float q2_pitch; // rad (elbow bend in same plane)
}

public struct FKResult
{
    public Vector3 basePos;
    public Vector3 joint1; // shoulder location (same as base here)
    public Vector3 joint2; // elbow
    public Vector3 ee;     // end effector
}

public struct IKResult
{
    public bool reachable;
    public JointAngles q;
    public string message;
}

public sealed class RobotArm
{
    private LinkParams _link1;
    private LinkParams _link2;

    public RobotArm(LinkParams l1, LinkParams l2)
    {
        _link1 = l1;
        _link2 = l2;
        _link1.RecomputeInertia();
        _link2.RecomputeInertia();
    }

    public float MaxReach() => _link1.length_m + _link2.length_m;
    public float MinReach() => MathF.Abs(_link1.length_m - _link2.length_m);

    public ref readonly LinkParams Link1() => ref _link1;
    public ref readonly LinkParams Link2() => ref _link2;

    public IKResult SolveIK(Vector3 target, bool elbowUp = false)
    {
        IKResult outp = new() { reachable = false, message = "" };

        // Workspace rule for this simulation: z must be non-negative.
        if (target.Z < 0.0f)
        {
            outp.message = "Invalid target: z must be >= 0";
            return outp;
        }

        double x = target.X;
        double y = target.Y;
        double z = target.Z;

        double d = Math.Sqrt(x * x + y * y + z * z);
        double rmin = MinReach();
        double rmax = MaxReach();

        if (d < rmin - 1e-9 || d > rmax + 1e-9)
        {
            outp.message = $"Target radius |p|={d:0.###} is outside [{rmin:0.###}, {rmax:0.###}]";
            return outp;
        }

        // Base yaw in XY plane.
        double q0 = 0.0;
        if (Math.Abs(x) > 1e-12 || Math.Abs(y) > 1e-12)
            q0 = Math.Atan2(y, x);

        // Reduce to planar IK in (r,z), where r = sqrt(x^2 + y^2).
        double r = Math.Sqrt(x * x + y * y);
        double L1 = _link1.length_m;
        double L2 = _link2.length_m;

        // Elbow angle (law of cosines).
        double c2 = (r * r + z * z - L1 * L1 - L2 * L2) / (2.0 * L1 * L2);
        c2 = Clamp(c2, -1.0, 1.0);

        double q2 = Math.Acos(c2);
        if (!elbowUp) q2 = -q2; // default branch (elbow-down)

        // Shoulder angle.
        double s2 = Math.Sin(q2);
        double k1 = L1 + L2 * Math.Cos(q2);
        double k2 = L2 * s2;

        double q1 = Math.Atan2(z, r) - Math.Atan2(k2, k1);

        outp.reachable = true;
        outp.q.q0_yaw = (float)q0;
        outp.q.q1_pitch = (float)q1;
        outp.q.q2_pitch = (float)q2;
        outp.message = "OK";
        return outp;
    }

    public FKResult ForwardKinematics(JointAngles q)
    {
        FKResult fk = new();
        fk.basePos = Vector3.Zero;
        fk.joint1 = fk.basePos;

        float L1 = _link1.length_m;
        float L2 = _link2.length_m;

        // u = radial direction in XY defined by yaw, k = +Z
        float cy = MathF.Cos(q.q0_yaw);
        float sy = MathF.Sin(q.q0_yaw);
        Vector3 u = new(cy, sy, 0.0f);
        Vector3 k = Vector3.UnitZ;

        // Elbow
        Vector3 p1 = u * (L1 * MathF.Cos(q.q1_pitch)) + k * (L1 * MathF.Sin(q.q1_pitch));

        // End effector
        float a = q.q1_pitch + q.q2_pitch;
        Vector3 p2 = p1 + u * (L2 * MathF.Cos(a)) + k * (L2 * MathF.Sin(a));

        fk.joint2 = p1;
        fk.ee = p2;
        return fk;
    }

    private static double Clamp(double v, double lo, double hi)
        => Math.Max(lo, Math.Min(hi, v));
}

File: Manipulator3D_Python/src/robot/robot_arm.py
from __future__ import annotations

from dataclasses import dataclass
from math import acos, atan2, cos, fabs, sin, sqrt
from typing import Tuple


Vec3 = Tuple[float, float, float]


@dataclass
class LinkParams:
    length_m: float = 2.5
    mass_kg: float = 1.0

    # Uniform rod approximations:
    # I_cm    = (1/12) m L^2   (about center, axis ⟂ to rod)
    # I_joint = (1/3)  m L^2   (about joint at one end)
    inertia_cm: float = 0.0
    inertia_joint: float = 0.0

    def recompute_inertia(self) -> None:
        self.inertia_cm = (1.0 / 12.0) * self.mass_kg * self.length_m * self.length_m
        self.inertia_joint = (1.0 / 3.0) * self.mass_kg * self.length_m * self.length_m


@dataclass
class JointAngles:
    q0_yaw: float = 0.0     # rad
    q1_pitch: float = 0.0   # rad (relative to XY plane)
    q2_pitch: float = 0.0   # rad (elbow relative, same vertical plane)


@dataclass
class FKResult:
    base: Vec3 = (0.0, 0.0, 0.0)
    joint1: Vec3 = (0.0, 0.0, 0.0)  # shoulder location (same as base)
    joint2: Vec3 = (0.0, 0.0, 0.0)  # elbow position
    ee: Vec3 = (0.0, 0.0, 0.0)      # end-effector position


@dataclass
class IKResult:
    reachable: bool
    q: JointAngles
    message: str


def _clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))


class RobotArm:
    def __init__(self, link1: LinkParams, link2: LinkParams) -> None:
        self._link1 = link1
        self._link2 = link2
        self._link1.recompute_inertia()
        self._link2.recompute_inertia()

    @property
    def link1(self) -> LinkParams:
        return self._link1

    @property
    def link2(self) -> LinkParams:
        return self._link2

    @property
    def l1(self) -> float:
        return self._link1.length_m

    @property
    def l2(self) -> float:
        return self._link2.length_m

    @property
    def max_reach(self) -> float:
        return self._link1.length_m + self._link2.length_m

    @property
    def min_reach(self) -> float:
        return fabs(self._link1.length_m - self._link2.length_m)

    def solve_ik(self, target: Vec3, elbow_up: bool = False) -> IKResult:
        x, y, z = target

        if z < 0.0:
            return IKResult(False, JointAngles(), "Invalid target: z must be >= 0")

        d = sqrt(x * x + y * y + z * z)
        rmin = self.min_reach
        rmax = self.max_reach
        if d < rmin - 1e-9 or d > rmax + 1e-9:
            return IKResult(
                False,
                JointAngles(),
                f"Target radius |p|={d:.6g} is outside [{rmin:.6g}, {rmax:.6g}]",
            )

        # Base yaw in XY
        q0 = 0.0
        if abs(x) > 1e-12 or abs(y) > 1e-12:
            q0 = atan2(y, x)

        # Reduce to planar IK in (r,z)
        r = sqrt(x * x + y * y)
        L1 = self._link1.length_m
        L2 = self._link2.length_m

        c2 = (r * r + z * z - L1 * L1 - L2 * L2) / (2.0 * L1 * L2)
        c2 = _clamp(c2, -1.0, 1.0)

        q2 = acos(c2)
        if not elbow_up:
            q2 = -q2

        s2 = sin(q2)
        k1 = L1 + L2 * cos(q2)
        k2 = L2 * s2

        q1 = atan2(z, r) - atan2(k2, k1)

        return IKResult(True, JointAngles(q0, q1, q2), "OK")

    def forward_kinematics(self, q: JointAngles) -> FKResult:
        L1 = self._link1.length_m
        L2 = self._link2.length_m

        cy = cos(q.q0_yaw)
        sy = sin(q.q0_yaw)

        # Radial unit in XY and +Z unit
        u = (cy, sy, 0.0)
        k = (0.0, 0.0, 1.0)

        # Elbow
        p1 = (
            u[0] * (L1 * cos(q.q1_pitch)) + k[0] * (L1 * sin(q.q1_pitch)),
            u[1] * (L1 * cos(q.q1_pitch)) + k[1] * (L1 * sin(q.q1_pitch)),
            u[2] * (L1 * cos(q.q1_pitch)) + k[2] * (L1 * sin(q.q1_pitch)),
        )

        # End effector
        a = q.q1_pitch + q.q2_pitch
        p2 = (
            p1[0] + u[0] * (L2 * cos(a)) + k[0] * (L2 * sin(a)),
            p1[1] + u[1] * (L2 * cos(a)) + k[1] * (L2 * sin(a)),
            p1[2] + u[2] * (L2 * cos(a)) + k[2] * (L2 * sin(a)),
        )

        return FKResult(base=(0.0, 0.0, 0.0), joint1=(0.0, 0.0, 0.0), joint2=p1, ee=p2)
